kernel/aster-nix/src/net/socket/unix/stream/listener.rs

```diff
@@ -144,42 +136,58 @@ impl BacklogTable {
 }
 
 struct Backlog {
+    addr: UnixSocketAddrBound,
     pollee: Pollee,
-    backlog: usize,
-    incoming_endpoints: Mutex<VecDeque<Endpoint>>,
+    backlog: AtomicUsize,
+    incoming_conns: Mutex<VecDeque<Connected>>,
 }
 
 impl Backlog {
-    fn new(backlog: usize) -> Self {
+    fn new(addr: UnixSocketAddrBound, backlog: usize) -> Self {
         Self {
+            addr,
             pollee: Pollee::new(IoEvents::empty()),
-            backlog,
-            incoming_endpoints: Mutex::new(VecDeque::with_capacity(backlog)),
+            backlog: AtomicUsize::new(backlog),
+            incoming_conns: Mutex::new(VecDeque::with_capacity(backlog)),
         }
     }
 
-    fn push_incoming(&self, endpoint: Endpoint) -> Result<()> {
-        let mut endpoints = self.incoming_endpoints.lock();
-        if endpoints.len() >= self.backlog {
-            return_errno_with_message!(Errno::ECONNREFUSED, "incoming_endpoints is full");
+    fn addr(&self) -> &UnixSocketAddrBound {
+        &self.addr
+    }
+
+    fn push_incoming(&self, client_addr: Option<UnixSocketAddrBound>) -> Result<Connected> {
+        let (server_conn, client_conn) = Connected::new_pair(Some(self.addr.clone()), client_addr);
+
+        let mut incoming_conns = self.incoming_conns.lock();
+        if incoming_conns.len() >= self.backlog.load(Ordering::Relaxed) {
+            return_errno_with_message!(
+                Errno::ECONNREFUSED,
+                "the pending connection queue on the listening socket is full"
+            );
         }
-        endpoints.push_back(endpoint);
+        incoming_conns.push_back(server_conn);
         self.pollee.add_events(IoEvents::IN);
-        Ok(())
+
+        Ok(client_conn)
     }
 
-    fn pop_incoming(&self) -> Option<Endpoint> {
-        let mut incoming_endpoints = self.incoming_endpoints.lock();
-        let endpoint = incoming_endpoints.pop_front();
-        if incoming_endpoints.is_empty() {
+    fn pop_incoming(&self) -> Result<Connected> {
+        let mut incoming_conns = self.incoming_conns.lock();
+        let conn = incoming_conns.pop_front();
+        if incoming_conns.is_empty() {
             self.pollee.del_events(IoEvents::IN);
         }
-        endpoint
+        conn.ok_or_else(|| Error::with_message(Errno::EAGAIN, "no pending connection is available"))
+    }
+
+    fn set_backlog(&self, backlog: usize) {
+        self.backlog.store(backlog, Ordering::Relaxed);
```

Comment:
If the new `backlog` is smaller than the number of incoming connections, perhaps we need to drop connections that exceed the new backlog. 

It could be acceptable to delay checking the actual effects and simply add a TODO or FIXME for now since setting new backlog is rare.

Reply:
Actually it seems Linux kernel implementation [won't drop the pending connections in this case](https://elixir.bootlin.com/linux/v6.10.5/source/net/unix/af_unix.c#L734).